### ISSUE TITLE
MLX Phase 1A: Router Contracts And Qwen3 Adapter

### ISSUE THREAD

**Issue by egesabanci** [egesabanci/reap-mlx#2]
## Objective

Define the MLX router output contract and implement the first architecture-specific router adapter for Qwen3-MoE using MLX-LM routing semantics.

## Background

Router correctness is the foundation of REAP saliency. The existing Torch observer derives selected experts from router logits using PyTorch APIs. The MLX path must not rely on generic `topk(softmax(logits))` for every architecture because supported MoE models route differently.

Qwen3-MoE is the natural first adapter because it has strong MLX-LM support and its routing behavior is relatively direct:

1. Flatten hidden states.
2. Apply `moe.gate`.
3. Full softmax over experts with `precise=True`.
4. Select top-k via `mx.argpartition`.
5. Gather selected scores with `mx.take_along_axis`.
6. Optionally renormalize selected scores if `norm_topk_prob` is true.

## Scope

Add the router data model and Qwen3 router adapter only.

Target files:

- `src/reap/backends/mlx/router.py`
- `tests/test_mlx_router.py`

## Implementation Steps

1. Create `src/reap/backends/mlx/router.py` without importing `torch` or `vllm`.
2. Add a `RouterResult` dataclass with fields:
   - `indices`: selected expert indices with shape `[batch, seq, top_k]` or `[tokens, top_k]` depending input rank.
   - `scores`: actual selected router scores with same leading shape as `indices`.
   - `logits`: optional raw logits, or `None` when not retained.
   - `score_mode`: string such as `"actual"` or future compatibility modes.
3. Implement `Qwen3MoeRouter`.
4. Prefer live module attributes over config values:
   - `mlp_layer.top_k` if present.
   - `mlp_layer.norm_topk_prob` if present.
   - fallback to config dict keys such as `num_experts_per_tok`, `top_k`, and `norm_topk_prob`.
5. Use `mx.argpartition` plus `mx.take_along_axis`; do not use `mx.topk` for indices.
6. Return actual scores selected from full softmax probabilities.
7. Preserve input rank: support `[batch, seq, hidden]` and `[tokens, hidden]`.
8. Add tests using tiny synthetic MLX modules or lightweight local stubs that do not download real models.

## Acceptance Criteria

- [ ] `RouterResult` is importable from `reap.backends.mlx.router`.
- [ ] `Qwen3MoeRouter` is importable from `reap.backends.mlx.router`.
- [ ] The router accepts both 2D and 3D hidden-state inputs.
- [ ] Returned indices are in `[0, num_experts)`.
- [ ] Returned scores have the same selected shape as indices.
- [ ] `norm_topk_prob=True` makes each token's selected scores sum to one.
- [ ] `norm_topk_prob=False` leaves selected full-softmax probabilities unrenormalized.
- [ ] Tests compare adapter output to a hand-written reference implementation matching Qwen3 MLX-LM routing.
- [ ] Importing the router module does not import `torch` or `vllm`.

## Test Plan

Run:

```bash
pytest tests/test_mlx_no_torch_import.py tests/test_mlx_router.py -v
```

Test cases to include:

- output shape for `[1, seq, hidden]`
- output shape for `[tokens, hidden]`
- valid expert-index range
- selected-score normalization when enabled
- exact or tolerance-based score parity against reference code
- edge case: one token
- edge case: `top_k == 1`

## Non-Goals

- Do not implement Mixtral in this issue.
- Do not implement observer replay.
- Do not implement pruning or save logic.
- Do not add PyTorch parity tests that require loading HF Torch models.
- Do not download large MLX-LM models in unit tests.

## References

- `implementation-plan-draft.md`, P1-001 and P1-002
- `docs/implementation-plan-v0.2.md`, Router Adapters
- `docs/reap-mlx-implementation-plan-gpt-v0.2.md`, Router Adapters

---

## Portfolio Objective Alignment - 2026-05-31

This issue is part of the Apple Silicon MLX REAP backend effort. The production/official REAP experimentation workflow remains the original PyTorch/CUDA implementation; this MLX work must not destabilize or replace that path.

The MLX backend goal is adapter-driven support for MoE weights, not a one-model port. Qwen3-specific work is bootstrap/reference-adapter work. The implementation should preserve REAP pruning semantics and observer-data schema compatibility while adapting to MLX constraints such as lazy execution, no PyTorch hooks, MLX-LM module layouts, and save/reload behavior.

For this issue, Qwen3 is only the first reference router adapter. The router contract must be general enough for future adapters to represent different routing semantics for arbitrary MoE weights.
